File: src/main/scala/pro/absolutne/scrapping/Browser.scala
package pro.absolutne.scrapping

import org.openqa.selenium.By._
import org.openqa.selenium.support.ui.ExpectedConditions._
import org.openqa.selenium.support.ui.{Select, WebDriverWait}
import org.openqa.selenium.{By, WebDriver, WebElement}
import scala.collection.JavaConverters._


class Browser(driver: WebDriver) {

  def goTo(url: String): Unit = driver.get(url)

  def find(css: String): WebElement = driver.findElement(cssSelector(css))

  def find(parent: WebElement, css: String): WebElement
  = parent.findElement(cssSelector(css))

  def tryFind(css: String): Option[WebElement] = {
    try {
      Some(find(css))
    } catch {
      case e:Exception => None
    }
  }

  def findAll(parent: WebElement, css: String): List[WebElement]
  = parent.findElements(cssSelector(css)).asScala.toList

  def findByText(el: WebElement, value: String): Unit = {
    new Select(el).selectByVisibleText(value)
  }

  def findVisible(css: String): WebElement =
    new WebDriverWait(driver, 10).until(visibilityOfElementLocated(cssSelector(css)))

  def getParrent(el: WebElement) = el.findElement(xpath(".."))

  def isPresent(css: String) = driver.findElements(cssSelector(css)).size > 0

  def hasChild(el: WebElement, by: By) = el.findElements(by).size > 0
}

File: src/main/java/pro/absolutne/webscrap/Browser.java
package pro.absolutne.webscrap;

import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.Collection;
import java.util.List;
import java.util.Optional;

import static java.util.stream.Collectors.toList;
import static org.openqa.selenium.By.cssSelector;
import static org.openqa.selenium.By.xpath;
import static org.openqa.selenium.support.ui.ExpectedConditions.visibilityOfElementLocated;

public class Browser {

    private final WebDriver driver;

    /**
     * Wait time in seconds
     */
    private final int waitTime = 10;

    public Browser(WebDriver driver) {
        this.driver = driver;
    }

    public void goTo(String url) {
        driver.get(url);
    }


    private List<BrowserElement> wrapAll(Collection<WebElement> els) {
        return els.stream().map(this::wrap).collect(toList());
    }

    public void selectMatching(WebElement parent, String value) {
        new Select(parent).selectByVisibleText(value);
    }

    public BrowserElement find(String sel) {
        return wrap(driver.findElement(cssSelector(sel)));
    }

    public BrowserElement findWithText(String sel, String text) {
        return findAll(sel).stream()
                .filter(el -> el.getText().equals(text))
                .findFirst().get();
    }

    public Optional<BrowserElement> tryFind(String sel) {
        try {
            return Optional.of(find(sel));
        } catch (NoSuchElementException e) {
            return Optional.empty();
        }
    }

    public List<BrowserElement> findAll(String sel) {
        return wrapAll(driver.findElements(cssSelector(sel)));
    }

    public BrowserElement findVisible(String sel) {
        By cssBy = cssSelector(sel);
        WebElement el = new WebDriverWait(driver, waitTime)
                .until(visibilityOfElementLocated(cssBy));

        return wrap(el);
    }


    public BrowserElement getParrent(WebElement el) {
        return wrap(el.findElement(xpath("..")));
    }

    public boolean isPresent(String sel) {
        return driver.findElements(cssSelector(sel)).size() > 0;
    }

    BrowserElement wrap(WebElement el) {
        return new BrowserElement(el, this);
    }

    BrowserElement find(WebElement parent, String sel) {
        return wrap(parent.findElement(cssSelector(sel)));
    }

    List<BrowserElement> findAll(WebElement parent, String sel) {
        return wrapAll(parent.findElements(cssSelector(sel)));
    }
}
